File: java/twelve-days/src/main/java/TwelveDays.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class TwelveDays {

    private static final String[] DAYS = {
        "first", "second", "third", "fourth", "fifth", "sixth",
        "seventh", "eighth", "ninth", "tenth", "eleventh", "twelfth"
    };

    private static final String[] GIFTS = {
        "a Partridge in a Pear Tree", "two Turtle Doves", "three French Hens",
        "four Calling Birds", "five Gold Rings", "six Geese-a-Laying",
        "seven Swans-a-Swimming", "eight Maids-a-Milking", "nine Ladies Dancing",
        "ten Lords-a-Leaping", "eleven Pipers Piping", "twelve Drummers Drumming"
    };

    String verse(int verseNumber) {
        StringBuilder verseBuilder = new StringBuilder();
        verseBuilder.append("On the ")
                    .append(DAYS[verseNumber - 1])
                    .append(" day of Christmas my true love gave to me: ");

        if (verseNumber == 1) {
            verseBuilder.append(GIFTS[0]).append(".");
        } else {
            for (int i = verseNumber - 1; i > 0; i--) {
                verseBuilder.append(GIFTS[i]).append(", ");
            }
            verseBuilder.append("and ").append(GIFTS[0]).append(".");
        }
        return verseBuilder.toString() + "\n";
    }

    String verses(int startVerse, int endVerse) {
        return IntStream.rangeClosed(startVerse, endVerse)
                        .mapToObj(this::verse)
                        .collect(Collectors.joining("\n"));
    }

    String sing() {
        return verses(1, 12);
    }
}

File: javascript/twelve-days/twelve-days.js
//
// This is only a SKELETON file for the 'Twelve Days' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const gifts = [
  'a Partridge in a Pear Tree',
  'two Turtle Doves',
  'three French Hens',
  'four Calling Birds',
  'five Gold Rings',
  'six Geese-a-Laying',
  'seven Swans-a-Swimming',
  'eight Maids-a-Milking',
  'nine Ladies Dancing',
  'ten Lords-a-Leaping',
  'eleven Pipers Piping',
  'twelve Drummers Drumming',
];

const ordinals = [
  'first', 'second', 'third', 'fourth', 'fifth', 'sixth',
  'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth'
];

const getVerse = (day) => {
  const dayIndex = day - 1;
  const dayOrdinal = ordinals[dayIndex];
  let verseGifts = '';

  // Build the list of gifts for the current day, in reverse order
  const currentDayGifts = gifts.slice(0, day).reverse();

  if (day === 1) {
    verseGifts = currentDayGifts[0];
  } else {
    // Join all gifts except the last one with commas
    verseGifts = currentDayGifts.slice(0, -1).join(', ');
    // Add the last gift (Partridge) with 'and'
    verseGifts += ', and ' + currentDayGifts[currentDayGifts.length - 1];
  }


  return `On the ${dayOrdinal} day of Christmas my true love gave to me: ${verseGifts}.\n`;
};

export const recite = (startVerse, endVerse) => {
  // If endVerse is not provided, default it to startVerse to handle single verse case
  const finalEndVerse = endVerse === undefined ? startVerse : endVerse;

  let result = '';
  for (let i = startVerse; i <= finalEndVerse; i++) {
    result += getVerse(i);
    if (i < finalEndVerse) {
      result += '\n'; // Add an extra newline between verses
    }
  }
  return result;
};
